### PR TITLE
convert sparkline api to one request - fixes archive

### LINKED ISSUE #18088: Multiple requests may archive the same range request at the same time
We noticed a while ago that multiple requests may launch archiving for the same period.

To reproduce this:

* Disable browser archiving & Disable browser archiving for a segment
  * eg set `[General]enable_browser_archiving_triggering = 0
enable_general_settings_admin = 0
browser_archiving_disabled_enforce = 1`
* Have a site with say 30 active goal 
* Have generally data for a few days 
* Trigger archiving using `./console core:archive`
* Then go the Goals Overview page and request a date range eg `2021-09-02,2021,09,29`

You might then notice that all does goals trigger the archiving. Eg you could execute a query like 

```sql
select count(*) from archive_numeric_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() and name like 'done%'; -- number of unique added archives in the last few minutes which may be pretty much roughly like the number of goals
select count(*) from archive_numeric_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() ;// how many archive entries were added overall

-- same for blob
select count(*) from archive_blob_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() ;
```

I'm meaning that each of these requests like in the screenshot seem to be launching the archiving (eg `index.php?forceView=1&viewDataTable=sparklines&module=Goals&action=get&idGoal=7&allow_multiple=0&only_summary=1&idSite=12&period=week&date=today&segment=&showtitle=1&random=1482`):

image

As I'm seeing this recently more often I think this was a regression somewhere between 4.3 and 4.4. I quickly looked through the archiving related changes in https://github.com/matomo-org/matomo/compare/4.3.0...4.4.1 and couldn't find anything so far that would explain this. This one peaked my interest but pretty sure it cannot be what caused it: https://github.com/matomo-org/matomo/pull/17657/files

Possibly the regression was in 4.2.0 because of https://github.com/matomo-org/matomo/pull/17379/files maybe

I wonder if this one maybe previously prevented that each request launches the archiving at the same time? Once we can reproduce the issue, we could maybe see if it works with that PR and wouldn't launch multiple archives.

Let me know if there are any issues reproducing this issue.

refs https://github.com/matomo-org/matomo/issues/17774

The goal is that the archiving is triggered only once for all these requests. Meaning below query should return `1`:

```sql
select count(*) from archive_numeric_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() and name like 'done%'; -- number of unique added archives in the last few minutes which may be pretty much roughly like the number of goals
```

### PR DESCRIPTION
**Pull request by peterhashair**
### Description:
Fixes: #18088. 
convert sparking API to one request - fixes archive 
Copy the logic from here : [https://github.com/matomo-org/matomo/blob/e3218471244d08039b1b84f1c907f1be0e018768/plugins/Ecommerce/Controller.php#L37](https://github.com/matomo-org/matomo/blob/e3218471244d08039b1b84f1c907f1be0e018768/plugins/Ecommerce/Controller.php#L37)

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by peterhashair** [comment]
@tsteur do you want to try that one, I think this is kind of working, reduce my CPU a little bit on my local on a one year period, still have failed some XML compare tests but should be fine, I will fix and tidy code more, once we confirm it reduce the load.

**Comment by tsteur** [comment]
@peterhashair let's maybe keep the changes for the lock separate and only apply the UI changes here. Be good to check out my comment in chat to make use of Visualisations as much as possible.

**Comment by peterhashair** [comment]
@tsteur right, I was stuck on here, so go for the template option. Let me convert to Visualisations. 

```php
 $view = ViewDataTableFactory::build(Sparklines::ID, $apiMethod, 'Goals.' . __METHOD__, $forceDefault = true);
 ```

**Comment by peterhashair** [comment]
@tsteur as we discussed. I used the goal API `getMetries` method instead of `get` it should remove the for loop improve the performance a little bit.

**Comment by tsteur** [comment]
Awesome, can you check if it loads faster now?

**Comment by peterhashair** [comment]
@tsteur yes, about 3-5 seconds faster compared to load period for 6 months on my local.

File: plugins/Goals/Controller.php [hunk 1]
```diff
@@ -219,6 +223,37 @@ public function getEvolutionGraph(array $columns = array(), $idGoal = false, arr
...
+            throw new \Exception("Invalid action name '$action' for '$module' plugin.");
+        }
+
+        $apiAction = $apiProxy->buildApiActionName($module, $action);
+        foreach ($goals as $goal) {
+            //load Visualisations Sparkline
+            $view = ViewDataTableFactory::build(Sparklines::ID, $apiAction, 'Goals.' . __METHOD__, true);
+            $view->requestConfig->request_parameters_to_modify['idGoal'] = $goal['idgoal'];
+            $view->requestConfig->request_parameters_to_modify['allow_multiple'] =(int)$goal['allow_multiple'];
+            $view->requestConfig->request_parameters_to_modify['only_summary'] =1;
```
Comment (bx80) [inline plugins/Goals/Controller.php]:
```suggestion
            $view->requestConfig->request_parameters_to_modify['allow_multiple'] = (int)$goal['allow_multiple'];
            $view->requestConfig->request_parameters_to_modify['only_summary'] = 1;
```

File: plugins/Goals/Pages.php [hunk 2]
```diff
@@ -68,11 +79,12 @@ public function createGoalsOverviewPage($goals)
...
             $config->setOrder(25);
             $config->setIsNotWidgetizable();
-            $config->addParameters(array('allow_multiple' => (int) $goal['allow_multiple'], 'only_summary' => '1'));
+            $config->addParameters(array('allow_multiple' => (int)$goal['allow_multiple'], 'only_summary' => '1'));
             $widgets[] = $config;
         }
 
-        $container = $this->createWidgetizableWidgetContainer('GoalsOverview', $subcategory, $widgets);
+
+
```
Comment (bx80) [inline plugins/Goals/Pages.php]:
```suggestion
```
```suggestion
```

**Review by bx80 — APPROVED** [review]
Works well for me and definitely a bit quicker. I've flagged a couple of very minor code tidies. :+1:

**Comment by sgiehl** [comment]
@peterhashair @bx80 This actually kind of broke the goals overview page. If you would have had a look at the UI test failures you would have been able to see that. See https://builds-artifacts.matomo.org/matomo-org/matomo/m-18088-fixes-archive/52262/GoalsPages_overview.png
Please try to be a bit more accurate on that. It's not only part of the development to fix failing tests that are related, but also part of the review to check if tests are failing and if updates to any tests and expected UI files were actually valid.
I will now revert this PR. Feel free to recreate it, once it fully works.

**Comment by peterhashair** [comment]
@sgiehl right let me fix that it. But I got a little confused here, it seems like the preview Test hides the revenue doesn't matter on which conditions (I mean is that on purpose or a regression), the new approach has actually loaded the revenue when `$isEcommerceEnabled`. I add can hide revenue sparkline like it was or update the screenshot. which approach do you want me to?

**Comment by sgiehl** [comment]
@peterhashair Actually it doesn't really matter if that would have been a regression or not. It was not part of the issue to change the content of the reports in any way. And it's actually not only the additional sparkline, also the headline misses `Goal` and the other sparklines show differently. In general it would for sure be fine to fix regressions if there are any, but we should avoid doing that silently in PRs without mentioning it anywhere.

**Pull request merged**